ffi/include/wirefilter.h

```diff
@@ -0,0 +1,129 @@
+#ifndef _WIREFILTER_H_
+#define _WIREFILTER_H_
+
+#include <stdlib.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct wirefilter_scheme wirefilter_scheme_t;
+typedef struct wirefilter_execution_context wirefilter_execution_context_t;
+typedef struct wirefilter_filter_ast wirefilter_filter_ast_t;
+typedef struct wirefilter_filter wirefilter_filter_t;
+
+typedef struct {
+    const unsigned char *data;
+    size_t length;
+} wirefilter_rust_allocated_str_t;
+
+typedef struct {
+    const unsigned char *data;
+    size_t length;
+} wirefilter_static_rust_allocated_str_t;
+
+typedef struct {
+    const unsigned char *data;
+    size_t length;
+} wirefilter_externally_allocated_str_t;
+
+typedef struct {
+    const unsigned char *data;
+    size_t length;
+} wirefilter_externally_allocated_byte_arr_t;
+
+typedef union {
+    union {
+        uint8_t success;
+        struct {
+            uint8_t _res1;
+            wirefilter_rust_allocated_str_t msg;
+        } err;
+        struct {
+            uint8_t _res2;
+            wirefilter_filter_ast_t *ast;
+        } ok;
+    };
+} wirefilter_parsing_result_t;
+
+typedef enum {
```

Comment:
Hmm, I know that this is what we used, but it doesn't actually match our Rust definitions and might break someday - corresponding `enum` is defined with `#[repr(u8)]`.

With C, there is no way to specify representation for an enum, but a somewhat common hack is:

```c
enum wirefilter_type {
  WIREFILTER_TYPE_IP,
  WIREFILTER_TYPE_BYTES,
  WIREFILTER_TYPE_INT,
  WIREFILTER_TYPE_BOOL,
};

typedef uint8_t wirefilter_type_t;
```

Reply:
Maybe we should consider using `repr(C)` instead? It would make the enum size bigger of a few bytes but might be better than this hack?

Reply:
Idk. Then it also affects alignment of every type that enum is in, which is probably fine for our current types, but doesn't feel good long-term...

Reply:
Also note that C++ doesn't have the same problem and allows to clearly define size of any type, so if we add C++ headers at some point, they won't need such hacks.

Reply:
I understand but this hack will also hides potential warnings/errors from the C compiler. I find it safer to use `repr(C)`

Reply:
Hmm what sort of warnings do you have in mind? Do you mean passing in numbers that are not among listed variants? If so, C doesn't warn about that even with `repr(C)` because it's allowed by the language to support bitflags and such :(

Reply:
More like switch case scenario

Reply:
Ah fair enough, didn't see that they do that nowadays and not just assume that enum is an arbitrary integer. I'm fine with changing to `repr(C)` for `Type`.

Reply:
With gcc uising the `-Wall` flag you'll get an error similar to:
```
test.c:12:2: warning: enumeration value 'TEST_B' not handled in switch [-Wswitch]
```

Reply:
I have modified the `repr` for `Type`.

Reply:
Yep, already checked myself with Clang, looks good. Thanks for pointing this out.